common/src/Renderer/ShaderManager.cpp

```diff
@@ -19,58 +19,57 @@
 
 #include "ShaderManager.h"
 
-#include "CollectionUtils.h"
 #include "IO/Path.h"
 #include "IO/SystemPaths.h"
 #include "Renderer/Shader.h"
 #include "Renderer/ShaderConfig.h"
+#include "Renderer/ShaderProgram.h"
+
+#include <cassert>
 
 namespace TrenchBroom {
     namespace Renderer {
-        ShaderManager::~ShaderManager() {
-            MapUtils::clearAndDelete(m_programs);
-            MapUtils::clearAndDelete(m_shaders);
-        }
+        ShaderManager::~ShaderManager() = default;
 
         ShaderProgram& ShaderManager::program(const ShaderConfig& config) {
-            ShaderProgramCache::iterator it = m_programs.find(&config);
-            if (it != std::end(m_programs))
+            auto it = m_programs.find(&config);
+            if (it != std::end(m_programs)) {
                 return *it->second;
+            }
 
-            ShaderProgram* program = createProgram(config);
-            m_programs.insert(ShaderProgramCacheEntry(&config, program));
-            return *program;
+            auto result = m_programs.emplace(&config, createProgram(config));
```

Comment:
this is weird that is stores a pointer to `config`, given that config is passed by const reference. I know it was like that before, I'd add a fixme to change the function signature to `program(const ShaderConfig* config)`

Reply:
Agreed, I'm changing it right away though.

Reply:
on second thought, it makes sense if you look at how we are using it; the shader configs are declared as constants in Shaders.h. It's a bit iffy, but if we take a pointer, we have to pass the address everywhere, which makes using the API odd. The fact that it's stored by pointer in the map seems like an implementation detail to me.